File: src/lib/logger.ts
/**
 * Structured Logger — Runtime-configurable log mode.
 *
 * Rules:
 * - error: Always shown (production + dev) — never suppressed, these are critical
 * - warn: Shown when logMode is enabled OR in __DEV__
 * - info / debug: Only shown when logMode is enabled (regardless of __DEV__)
 *
 * Log mode is toggled via Settings → Developer Tools. Turning on devMode
 * automatically enables logMode, but logMode can be turned off independently.
 */

let _logModeEnabled = __DEV__;

/** Enable or disable verbose logging at runtime */
export function setLogMode(enabled: boolean): void {
    _logModeEnabled = enabled;
}

/** Check if verbose logging is currently enabled */
export function getLogMode(): boolean {
    return _logModeEnabled;
}

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

/** Central log function. Prefer the tagged helpers below. */
export function logger(level: LogLevel, tag: string, message: string, ...args: unknown[]): void {
    const prefix = `[${tag}] ${message}`;
    if (level === 'error') {
        console.error(prefix, ...args);
        return;
    }
    if (!_logModeEnabled) return;
    if (level === 'warn') {
        console.warn(prefix, ...args);
    } else {
        // eslint-disable-next-line no-console
        console.log(prefix, ...args);
    }
}

/** Tagged helpers for common domains */
export const logStorage = (level: LogLevel, message: string, ...args: unknown[]) => logger(level, 'Storage', message, ...args);
export const logAi = (level: LogLevel, message: string, ...args: unknown[]) => logger(level, 'AI', message, ...args);
export const logAiQueue = (level: LogLevel, message: string, ...args: unknown[]) => logger(level, 'AI Queue', message, ...args);
export const logCompressor = (level: LogLevel, message: string, ...args: unknown[]) => logger(level, 'Compressor', message, ...args);
export const logDb = (level: LogLevel, message: string, ...args: unknown[]) => logger(level, 'DB', message, ...args);
export const logStartup = (level: LogLevel, message: string, ...args: unknown[]) => logger(level, 'Startup', message, ...args);

File: mda_flutter/lib/core/logger.dart
/// Minimal logger ported from `src/lib/logger.ts`.
/// Levels: error always shown; warn only in dev/logMode; info/debug only in logMode.
library;

import 'package:flutter/foundation.dart';

bool _logMode = kDebugMode;

/// Sets verbose logging mode (settings pref `LOG_MODE`).
void setLogMode(bool enabled) => _logMode = enabled;

bool getLogMode() => _logMode;

enum LogLevel { debug, info, warn, error }

void _log(LogLevel level, String tag, String message, [Object? extra]) {
  final show = level == LogLevel.error ||
      (level == LogLevel.warn && (_logMode || kDebugMode)) ||
      (level != LogLevel.error && _logMode);
  if (!show) return;
  final prefix = switch (level) {
    LogLevel.error => '❌',
    LogLevel.warn => '⚠️',
    LogLevel.info => 'ℹ️',
    LogLevel.debug => '🐞',
  };
  // ignore: avoid_print
  print('$prefix [$tag] $message${extra != null ? ' $extra' : ''}');
}

/// Generic logger.
void logger(LogLevel level, String tag, String message, [Object? extra]) =>
    _log(level, tag, message, extra);

/// Tagged helpers matching the RN logger.
final logStorage = _tagged('STORAGE');
final logAi = _tagged('AI');
final logAiQueue = _tagged('AI_QUEUE');
final logCompressor = _tagged('COMPRESSOR');
final logDb = _tagged('DB');
final logStartup = _tagged('STARTUP');
final logVlog = _tagged('VLOG');

LogFunction _tagged(String tag) => LogFunction(tag);

/// A small callable wrapper so `logDb.info('msg')` reads like the RN API.
class LogFunction {
  LogFunction(this.tag);

  final String tag;

  void error(String message, [Object? extra]) => _log(LogLevel.error, tag, message, extra);
  void warn(String message, [Object? extra]) => _log(LogLevel.warn, tag, message, extra);
  void info(String message, [Object? extra]) => _log(LogLevel.info, tag, message, extra);
  void debug(String message, [Object? extra]) => _log(LogLevel.debug, tag, message, extra);
}
